### ISSUE TITLE
ArgumentException: The umbraco domain could not be interpreted

### ISSUE THREAD

**Issue by weinerm** [Infocaster/UrlTracker#53]
**Describe the bug**
_Using version: 9.0.1_

When creating a redirect, some domains are missing from the "Root node" dropdown.
Only actual root nodes are shown, however non-root nodes that are configured with a hostname are not shown. I have a unique situation why I do not want some domains as root notes.

**To Reproduce**
1. Add a root node
2. Add a node under the root node
3. Assign a hostname to the child node
4. Attempt to add a redirect for the child node

**Expected behavior**
All nodes with specified hostnames should appear in the Root node dropdown. This was the behavior in v3.11

**Comment by D-Inventor** [comment]
Hi again :D

Just to be sure, have you enabled the option `HasDomainOnChildNode` in the configuration?

You can [read a little bit more about this option in the documentation](https://github.com/Infocaster/UrlTracker/wiki/Configuration#hasdomainonchildnode).  
Read more about setting configurations in [the first paragraph](https://github.com/Infocaster/UrlTracker/wiki/Configuration#version-9--10).

**Comment by weinerm** [comment]
Hello!

Who would have thought it would actually be helpful reading the documentation! Sigh, sorry about that.

Anyways, I added the configuration within appsettings.json, but am now getting this error running via IIS Express locally. I do have a node configured with the localhost:port hostname.

```
An unhandled exception occurred while processing the request.
ArgumentException: The umbraco domain could not be interpreted. (Parameter 'domain')
UrlTracker.Core.Domain.DomainProvider.GetUrlFromDomain(IDomain domain, UrlTrackerSettings settings, IUmbracoContextReferenceAbstraction cref)
```

**Comment by D-Inventor** [comment]
That's interesting, I'm gonna try to reproduce the issue and get back to you if I've discovered anything!

**Comment by D-Inventor** [comment]
@weinerm this is my analysis thus far:

This is the part of the source that throws the exception:
```csharp
// I'm taking some liberties on the original logic. The old logic ensured that the url starts with a protocol.
//    I am parsing the url into an object to make comparisons with incoming requests easier and more straight forward
string? result = null;

// This condition is taken from the old logic. I'm not sure why it works like this...
if (settings.HasDomainOnChildNode && domain.RootContentId.HasValue)
{
    var node = cref.GetContentById(domain.RootContentId.Value);
    if (node is not null && node.Level > 1)
    {
        result = node.Url(_umbracoContextFactory, mode: UrlMode.Absolute);
    }
}
else
{
    // otherwise read it immediately from the domain
    result = domain.DomainName;
}

return Url.Parse(result ?? throw new ArgumentException("The umbraco domain could not be interpreted.", nameof(domain)));
```
[source](https://github.com/Infocaster/UrlTracker/blob/v10/develop/src/UrlTracker.Core/Domain/DomainProvider.cs#L67)

The only way for you to trigger this exception, would be to have the `HasDomainOnChildNode` be true, have a domain with a root content id, but either being unable to get the node (that is: `node is null`) or the node is on level 1 or lower.

Now as the comment says: I am not sure why it works like this, as this logic has been taken from the old 4.x version before I completely rewrote the backend. Do any of these conditions apply to your situation?

**Comment by weinerm** [comment]
I've tried to summarize a simplified version of my multi-site set up. Which will hopefully help you diagnose/reproduce.

Domain 1 (root)
Domain 2 (root)
Domain 3 (root)
-- Domain 4 (level 1)
-- Domain 5 (level 1)

**Comment by D-Inventor** [comment]
That confirms what I thought!

I shared the snippet with my coworkers to make sure I didn't overlook anything and I think I know what to change to solve this issue. I'm gonna make a code change and come back so we can try it out and see if it solves your issue!

**Comment by D-Inventor** [comment]
@weinerm do you have an opportunity to try out [version 9.0.2-beta0001](https://www.nuget.org/packages/UrlTracker/9.0.2-beta0001)?
That version has a fix for this issue

**Comment by weinerm** [comment]
I'm out at a work conference for a few days. Will get back to you next week. Thanks!

**Comment by weinerm** [comment]
@D-Inventor I no longer get an exception with HasDomainOnChildNode set to true on version 9.0.2-beta0001
However, back to my original issue, it only lets me select "root" nodes in the drop-down when creating a redirect. I have several Level-1 nodes that have hostnames assigned that are not showing in the drop-down.

**Comment by D-Inventor** [comment]
I'll have a look to see if I can find anything

**Comment by D-Inventor** [comment]
Looks like it's going wrong in the frontend [over here](https://github.com/Infocaster/UrlTracker/blob/v9/main/src/UrlTracker.Web/content/App_Plugins/UrlTracker/Scripts/UrlTracker.DetailsController.js#L111).

I had hoped I wouldn't have to touch the frontend until the redesign, but there's no way around it, it seems. I'll get back to you when I have a fix, which will probably be in a few days.

**Comment by D-Inventor** [comment]
[Version 9.0.2-beta0003](https://www.nuget.org/packages/UrlTracker/9.0.2-beta0003) Should fix this issue. Do you have a chance to try it out?

**Comment by weinerm** [comment]
Confirmed fixed. Nice work and thank you!

**Comment by emilybotha** [comment]
Hi there,

We're having this exact issue in our Umbraco 8.12.1 setup where we have our site domains set on child nodes and not the top-level node so URL Tracker isn't working for us at all right now.

Is there any way to fix this issue in URL Tracker v4.1.132? We're stuck on our current version of U8 because of some breaking changes that will affect us in the later versions but we can't upgrade to a newer version of URL Tracker without upgrading Umbraco 8.

Any help would be much appreciated. 

Thanks,
Em

**Comment by D-Inventor** [comment]
Hi Em! Thank you for reaching out! I would much prefer if you update to v5 of the URL Tracker. How about I'll lower the dependency version on v5, so that you can install it as well?

That being said: the fixes and changes from v9 and v10 are yet to be ported back to v5. This takes a bit longer with the change in framework, but I'll get onto it. Let's make this a separate issue.
